File: Apps/HIL/android/app/src/main/java/com/dweekly/cyrinxhil/ReverseBurstCodec.kt
package com.dweekly.cyrinxhil

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.sin
import kotlin.math.sqrt

class ReverseBurstCodec(config: SessionConfig) {
    private val sampleRateHz = config.sampleRateHz.coerceIn(8_000, 192_000)
    private val lowHz = minOf(config.bandStartHz, config.bandEndHz)
    private val highHz = maxOf(config.bandStartHz, config.bandEndHz)
    private val symbolSamples = config.dcssSymbolSamples.coerceIn(240, 4096)
    private val txGainCap = config.txGainCap.coerceIn(0.01f, 0.80f)
    private val lowRef = makeReference(lowHz)
    private val highRef = makeReference(highHz)
    private val leaderSymbols = 6
    private val gapSymbols = 1
    private val minLeaderRms = 0.0025f
    private val symbolSearchRadius = max(8, symbolSamples / 8)
    private val symbolSearchStep = max(1, symbolSamples / 48)
    private val guardBytes = intArrayOf(0x55, 0x2D)
    private val fixedPayloadBytes = 5
    private val repetitionCount = 5
    private val preambleBits = IntArray(24) { idx -> if ((idx % 2) == 0) 1 else 0 } + byteToBits(0x7E)
    private val minPreambleScore = preambleBits.size - 4
    private val rxBuffer = ArrayList<Float>()

    fun encode(payload: ByteArray): FloatArray {
        val trimmed = payload.copyOfRange(0, minOf(payload.size, fixedPayloadBytes))
        val padded = ByteArray(fixedPayloadBytes)
        System.arraycopy(trimmed, 0, padded, 0, trimmed.size)
        val length = trimmed.size and 0xFF
        val crcInput = ByteArray(fixedPayloadBytes + 1)
        System.arraycopy(padded, 0, crcInput, 0, fixedPayloadBytes)
        crcInput[fixedPayloadBytes] = length.toByte()
        val frame = ArrayList<Byte>(guardBytes.size + ((fixedPayloadBytes + 2) * repetitionCount))
        frame.add(guardBytes[0].toByte())
        frame.add(guardBytes[1].toByte())
        for (value in padded) {
            repeat(repetitionCount) { frame.add(value) }
        }
        repeat(repetitionCount) { frame.add(length.toByte()) }
        val crc = crc8(crcInput, 0, crcInput.size).toByte()
        repeat(repetitionCount) { frame.add(crc) }

        val bits = ArrayList<Int>(preambleBits.size + (frame.size * 8))
        for (bit in preambleBits) {
            bits.add(bit)
        }
        for (byte in frame) {
            for (bit in byteToBits(byte.toInt() and 0xFF)) {
                bits.add(bit)
            }
        }

        val totalSamples = (leaderSymbols + gapSymbols + bits.size) * symbolSamples
        val out = FloatArray(totalSamples)
        var cursor = 0
        repeat(leaderSymbols) {
            System.arraycopy(lowRef, 0, out, cursor, symbolSamples)
            cursor += symbolSamples
        }
        cursor += gapSymbols * symbolSamples
        for (bit in bits) {
            val ref = if (bit == 0) lowRef else highRef
            System.arraycopy(ref, 0, out, cursor, symbolSamples)
            cursor += symbolSamples
        }
        return out
    }

    fun ingest(samples: FloatArray): List<ByteArray> {
        if (samples.isEmpty()) {
            return emptyList()
        }
        for (sample in samples) {
            rxBuffer.add(sample)
        }
        return decodeAvailable()
    }

    private fun decodeAvailable(): List<ByteArray> {
        val decoded = ArrayList<ByteArray>()
        val frameBytes = guardBytes.size + ((fixedPayloadBytes + 2) * repetitionCount)
        val totalBits = preambleBits.size + (frameBytes * 8)
        val minSamples = (leaderSymbols + gapSymbols + totalBits) * symbolSamples

        while (rxBuffer.size >= minSamples) {
            val leaderStart = findLeaderStart() ?: break
            val nominalStart = leaderStart + ((leaderSymbols + gapSymbols) * symbolSamples)
            val refinedStart = refinePayloadStart(nominalStart)
            if (refinedStart == null) {
                dropFront(leaderStart + symbolSamples)
                continue
            }
            val searchFrom = max(0, refinedStart - (symbolSamples / 2))
            val searchTo = minOf(rxBuffer.size - (totalBits * symbolSamples), refinedStart + (symbolSamples / 2))
            var acceptedPayload: ByteArray? = null
            var acceptedFrameEnd = -1
            if (searchFrom <= searchTo) {
                for (candidateStart in searchFrom..searchTo) {
                    val preambleScore = scorePreambleAt(candidateStart)
                    if (preambleScore < minPreambleScore - 1) {
                        continue
                    }
                    val frame = ByteArray(frameBytes)
                    var valid = true
                    for (index in 0 until frameBytes) {
                        val value = decodeByteAt(candidateStart, preambleBits.size + (index * 8))
                        if (value == null) {
                            valid = false
                            break
                        }
                        frame[index] = value.toByte()
                    }
                    if (!valid) {
                        continue
                    }

                    val payloadStart = guardBytes.size
                    val correctedPayload = ByteArray(fixedPayloadBytes)
                    for (index in 0 until fixedPayloadBytes) {
                        val base = payloadStart + (index * repetitionCount)
                        correctedPayload[index] = majorityByte(frame.copyOfRange(base, base + repetitionCount))
                    }
                    val lengthBase = payloadStart + (fixedPayloadBytes * repetitionCount)
                    val length = (majorityByte(frame.copyOfRange(lengthBase, lengthBase + repetitionCount)).toInt() and 0xFF).coerceAtMost(fixedPayloadBytes)
                    val crcInput = ByteArray(fixedPayloadBytes + 1)
                    System.arraycopy(correctedPayload, 0, crcInput, 0, fixedPayloadBytes)
                    crcInput[fixedPayloadBytes] = length.toByte()
                    val expectedCrc = crc8(crcInput, 0, crcInput.size)
                    val crcBase = lengthBase + repetitionCount
                    val actualCrc = majorityByte(frame.copyOfRange(crcBase, crcBase + repetitionCount)).toInt() and 0xFF
                    if (expectedCrc != actualCrc) {
                        continue
                    }

                    acceptedPayload = correctedPayload.copyOfRange(0, length)
                    acceptedFrameEnd = candidateStart + (totalBits * symbolSamples)
                    break
                }
            }

            if (acceptedPayload == null) {
                dropFront(leaderStart + symbolSamples)
                continue
            }

            decoded.add(acceptedPayload)
            dropFront(acceptedFrameEnd)
        }

        val keep = max(symbolSamples * 16, minSamples * 2)
        if (rxBuffer.size > keep) {
            dropFront(rxBuffer.size - keep)
        }
        return decoded
    }

    private fun findLeaderStart(): Int? {
        val maxStart = rxBuffer.size - ((leaderSymbols + gapSymbols + preambleBits.size) * symbolSamples)
        if (maxStart <= 0) {
            return null
        }
        var bestStart = -1
        var bestScore = Int.MIN_VALUE
        val coarseStep = max(8, symbolSamples / 6)
        var coarse = 0
        while (coarse <= maxStart) {
            val score = scoreLeaderAt(coarse)
            if (score > bestScore) {
                bestScore = score
                bestStart = coarse
            }
            coarse += coarseStep
        }
        if (bestStart < 0) {
            return null
        }
        val refineFrom = max(0, bestStart - coarseStep)
        val refineTo = minOf(maxStart, bestStart + coarseStep)
        for (candidate in refineFrom..refineTo) {
            val score = scoreLeaderAt(candidate)
            if (score > bestScore) {
                bestScore = score
                bestStart = candidate
            }
        }
        return if (bestScore >= leaderSymbols) bestStart else null
    }

    private fun scoreLeaderAt(start: Int): Int {
        var score = 0
        for (symbolIndex in 0 until leaderSymbols) {
            val windowStart = start + (symbolIndex * symbolSamples)
            if (windowRms(windowStart) < minLeaderRms) {
                return Int.MIN_VALUE
            }
            val bit = decodeBitAt(start, symbolIndex) ?: return Int.MIN_VALUE
            if (bit == 0) {
                score += 1
            }
        }
        return score
    }

    private fun scorePreambleAt(start: Int): Int {
        var score = 0
        for (bitIndex in preambleBits.indices) {
            val bit = decodeBitAt(start, bitIndex) ?: return Int.MIN_VALUE
            if (bit == preambleBits[bitIndex]) {
                score += 1
            }
        }
        return score
    }

    private fun refinePayloadStart(nominalStart: Int): Int? {
        val maxStart = rxBuffer.size - (preambleBits.size * symbolSamples)
        if (maxStart <= 0) {
            return null
        }
        val searchRadius = symbolSamples
        val from = max(0, nominalStart - searchRadius)
        val to = minOf(maxStart, nominalStart + searchRadius)
        if (from > to) {
            return null
        }
        var bestStart = -1
        var bestScore = Int.MIN_VALUE
        for (candidate in from..to) {
            val score = scorePreambleAt(candidate)
            if (score > bestScore) {
                bestScore = score
                bestStart = candidate
            }
        }
        return if (bestScore >= minPreambleScore) bestStart else null
    }

    private fun decodeByteAt(startSample: Int, startBit: Int): Int? {
        var value = 0
        for (bitOffset in 0 until 8) {
            val bit = decodeBitAt(startSample, startBit + bitOffset) ?: return null
            value = (value shl 1) or bit
        }
        return value and 0xFF
    }

    private fun decodeBitAt(startSample: Int, bitIndex: Int): Int? {
        val nominalStart = startSample + (bitIndex * symbolSamples)
        if (nominalStart < 0) {
            return null
        }
        val searchFrom = max(0, nominalStart - symbolSearchRadius)
        val searchTo = minOf(rxBuffer.size - symbolSamples, nominalStart + symbolSearchRadius)
        if (searchFrom > searchTo) {
            return null
        }
        var bestBit: Int? = null
        var bestScore = Float.NEGATIVE_INFINITY
        var candidateStart = searchFrom
        while (candidateStart <= searchTo) {
            val result = classifyBit(candidateStart)
            val strength = abs(result.second - result.third)
            if (strength > bestScore) {
                bestScore = strength
                bestBit = result.first
            }
            candidateStart += symbolSearchStep
        }
        return bestBit
    }

    private fun classifyBit(windowStart: Int): Triple<Int, Float, Float> {
        var lowI = 0f
        var lowQ = 0f
        var highI = 0f
        var highQ = 0f
        for (idx in 0 until symbolSamples) {
            val sample = rxBuffer[windowStart + idx]
            lowI += sample * lowRef[idx]
            lowQ += sample * lowRef[idx + symbolSamples]
            highI += sample * highRef[idx]
            highQ += sample * highRef[idx + symbolSamples]
        }
        val lowMag = sqrt((lowI * lowI) + (lowQ * lowQ))
        val highMag = sqrt((highI * highI) + (highQ * highQ))
        return Triple(if (highMag > lowMag) 1 else 0, lowMag, highMag)
    }

    private fun makeReference(freqHz: Int): FloatArray {
        val out = FloatArray(symbolSamples * 2)
        for (idx in 0 until symbolSamples) {
            val phase = (2.0 * PI.toFloat() * freqHz.toFloat() * idx.toFloat()) / sampleRateHz.toFloat()
            out[idx] = txGainCap * sin(phase.toDouble()).toFloat()
            out[idx + symbolSamples] = txGainCap * cos(phase.toDouble()).toFloat()
        }
        return out
    }

    private fun byteToBits(value: Int): IntArray {
        return IntArray(8) { bitIndex -> (value ushr (7 - bitIndex)) and 0x1 }
    }

    private fun majorityByte(bytes: ByteArray): Byte {
        var value = 0
        for (bitIndex in 0 until 8) {
            val mask = 1 shl (7 - bitIndex)
            var ones = 0
            for (byte in bytes) {
                if ((byte.toInt() and mask) != 0) {
                    ones += 1
                }
            }
            if ((ones * 2) >= bytes.size) {
                value = value or mask
            }
        }
        return value.toByte()
    }

    private fun crc8(bytes: ByteArray, start: Int, length: Int): Int {
        var crc = 0
        for (index in 0 until length) {
            crc = crc xor (bytes[start + index].toInt() and 0xFF)
        }
        return crc and 0xFF
    }

    private fun dropFront(count: Int) {
        if (count <= 0 || rxBuffer.isEmpty()) {
            return
        }
        rxBuffer.subList(0, minOf(count, rxBuffer.size)).clear()
    }

    private fun windowRms(start: Int): Float {
        val end = start + symbolSamples
        if (start < 0 || end > rxBuffer.size) {
            return 0f
        }
        var energy = 0f
        for (index in start until end) {
            val sample = rxBuffer[index]
            energy += sample * sample
        }
        return sqrt(energy / symbolSamples.toFloat())
    }
}

File: Sources/Cyrinx/ReverseBurstCodec.swift
import Foundation

struct ReverseBurstCodec {
    private let sampleRateHz: Int
    private let lowHz: Int
    private let highHz: Int
    private let symbolSamples: Int
    private let txGainCap: Float
    private let lowRef: (sin: [Float], cos: [Float])
    private let highRef: (sin: [Float], cos: [Float])
    private let leaderSymbols = 6
    private let gapSymbols = 1
    private let minLeaderRms: Float = 0.0025
    private let symbolSearchRadius: Int
    private let symbolSearchStep: Int
    private let preambleBits: [UInt8]
    private let minPreambleScore: Int
    private let guardBytes: [UInt8] = [0x55, 0x2D]
    private let fixedPayloadBytes = 5
    private let repetitionCount = 5
    private var rxBuffer: [Float] = []

    init(config: Config, symbolSamples: Int) {
        sampleRateHz = Int(min(max(config.sampleRateHz, 8_000), 192_000))
        lowHz = Int(min(config.bandStartHz, config.bandEndHz))
        highHz = Int(max(config.bandStartHz, config.bandEndHz))
        self.symbolSamples = min(max(symbolSamples, 240), 4096)
        txGainCap = min(max(config.txGainCap, 0.01), 0.80)
        lowRef = Self.makeReference(
            freqHz: lowHz,
            sampleRateHz: sampleRateHz,
            symbolSamples: self.symbolSamples,
            txGainCap: txGainCap
        )
        highRef = Self.makeReference(
            freqHz: highHz,
            sampleRateHz: sampleRateHz,
            symbolSamples: self.symbolSamples,
            txGainCap: txGainCap
        )
        symbolSearchRadius = max(8, self.symbolSamples / 8)
        symbolSearchStep = max(1, self.symbolSamples / 48)
        let preamble = (0..<24).map { UInt8(($0 % 2) == 0 ? 1 : 0) }
        preambleBits = preamble + Self.byteToBits(0x7E)
        minPreambleScore = preambleBits.count - 4
    }

    mutating func encode(payload: Data) -> [Float] {
        let trimmed = payload.prefix(fixedPayloadBytes)
        var padded = [UInt8](repeating: 0, count: fixedPayloadBytes)
        for (index, byte) in trimmed.enumerated() {
            padded[index] = byte
        }
        let length = UInt8(trimmed.count)
        let crc = UInt8(Self.crc8(Array(padded) + [length]))
        var frameBytes = guardBytes
        for byte in padded {
            frameBytes.append(contentsOf: Array(repeating: byte, count: repetitionCount))
        }
        frameBytes.append(contentsOf: Array(repeating: length, count: repetitionCount))
        frameBytes.append(contentsOf: Array(repeating: crc, count: repetitionCount))

        var bits = preambleBits
        for byte in frameBytes {
            bits.append(contentsOf: Self.byteToBits(byte))
        }

        let totalSymbols = leaderSymbols + gapSymbols + bits.count
        var output = [Float](repeating: 0, count: totalSymbols * symbolSamples)
        var cursor = 0
        for _ in 0..<leaderSymbols {
            writeWaveform(lowRef.sin, into: &output, offset: cursor)
            cursor += symbolSamples
        }
        cursor += gapSymbols * symbolSamples
        for bit in bits {
            let ref = bit == 0 ? lowRef.sin : highRef.sin
            writeWaveform(ref, into: &output, offset: cursor)
            cursor += symbolSamples
        }
        return output
    }

    mutating func ingest(samples: UnsafeBufferPointer<Float>) -> [Data] {
        if samples.isEmpty {
            return []
        }
        rxBuffer.append(contentsOf: samples)
        return decodeAvailable()
    }

    mutating func decodeWaveform(_ samples: [Float]) -> [Data] {
        samples.withUnsafeBufferPointer { ingest(samples: $0) }
    }

    private mutating func decodeAvailable() -> [Data] {
        var decoded: [Data] = []
        let frameBytes = guardBytes.count + ((fixedPayloadBytes + 2) * repetitionCount)
        let totalBits = preambleBits.count + (frameBytes * 8)
        let minSamples = (leaderSymbols + gapSymbols + totalBits) * symbolSamples

        while rxBuffer.count >= minSamples {
            guard let leaderStart = findLeaderStart() else {
                break
            }
            let nominalStart = leaderStart + ((leaderSymbols + gapSymbols) * symbolSamples)
            guard let refinedStart = refinePayloadStart(near: nominalStart) else {
                dropFront(leaderStart + symbolSamples)
                continue
            }

            let searchFrom = max(0, refinedStart - (symbolSamples / 2))
            let searchTo = min(rxBuffer.count - (totalBits * symbolSamples), refinedStart + (symbolSamples / 2))
            var accepted: (payload: Data, frameEnd: Int)?
            if searchFrom <= searchTo {
                for candidateStart in searchFrom...searchTo {
                    let preambleScore = scorePreamble(at: candidateStart)
                    if preambleScore < minPreambleScore - 1 {
                        continue
                    }
                    var frame = [UInt8](repeating: 0, count: frameBytes)
                    var valid = true
                    for index in 0..<frameBytes {
                        guard let byte = decodeByte(at: candidateStart, startBit: preambleBits.count + (index * 8)) else {
                            valid = false
                            break
                        }
                        frame[index] = byte
                    }
                    if !valid {
                        continue
                    }

                    let payloadTriples = frame[guardBytes.count..<(guardBytes.count + (fixedPayloadBytes * repetitionCount))]
                    var correctedPayload = [UInt8]()
                    correctedPayload.reserveCapacity(fixedPayloadBytes)
                    for index in 0..<fixedPayloadBytes {
                        let start = payloadTriples.startIndex + (index * repetitionCount)
                        let end = start + repetitionCount
                        correctedPayload.append(majorityByte(Array(payloadTriples[start..<end])))
                    }

                    let lengthBase = guardBytes.count + (fixedPayloadBytes * repetitionCount)
                    let length = Int(majorityByte(Array(frame[lengthBase..<(lengthBase + repetitionCount)])))
                    if length > fixedPayloadBytes {
                        continue
                    }

                    let crcBase = lengthBase + repetitionCount
                    let actual = Int(majorityByte(Array(frame[crcBase..<(crcBase + repetitionCount)])))
                    let expected = Self.crc8(correctedPayload + [UInt8(length)])
                    if expected != actual {
                        continue
                    }

                    accepted = (
                        Data(correctedPayload[0..<length]),
                        candidateStart + (totalBits * symbolSamples)
                    )
                    break
                }
            }

            guard let accepted else {
                dropFront(leaderStart + symbolSamples)
                continue
            }
            decoded.append(accepted.payload)
            dropFront(accepted.frameEnd)
        }

        let keep = max(symbolSamples * 16, minSamples * 2)
        if rxBuffer.count > keep {
            dropFront(rxBuffer.count - keep)
        }
        return decoded
    }

    private func findLeaderStart() -> Int? {
        let maxStart = rxBuffer.count - ((leaderSymbols + gapSymbols + preambleBits.count) * symbolSamples)
        if maxStart <= 0 {
            return nil
        }

        var bestStart = -1
        var bestScore = Int.min
        let coarseStep = max(8, symbolSamples / 6)
        var coarse = 0
        while coarse <= maxStart {
            let score = scoreLeader(at: coarse)
            if score > bestScore {
                bestScore = score
                bestStart = coarse
            }
            coarse += coarseStep
        }

        if bestStart < 0 {
            return nil
        }

        let refineFrom = max(0, bestStart - coarseStep)
        let refineTo = min(maxStart, bestStart + coarseStep)
        for candidate in refineFrom...refineTo {
            let score = scoreLeader(at: candidate)
            if score > bestScore {
                bestScore = score
                bestStart = candidate
            }
        }
        return bestScore >= leaderSymbols ? bestStart : nil
    }

    private func scoreLeader(at start: Int) -> Int {
        var score = 0
        for symbolIndex in 0..<leaderSymbols {
            let windowStart = start + (symbolIndex * symbolSamples)
            if windowRms(start: windowStart) < minLeaderRms {
                return Int.min
            }
            guard let bit = decodeBit(at: start, bitIndex: symbolIndex) else {
                return Int.min
            }
            if bit == 0 {
                score += 1
            }
        }
        return score
    }

    private func scorePreamble(at start: Int) -> Int {
        var score = 0
        for (index, expectedBit) in preambleBits.enumerated() {
            guard let bit = decodeBit(at: start, bitIndex: index) else {
                return Int.min
            }
            if bit == expectedBit {
                score += 1
            }
        }
        return score
    }

    private func refinePayloadStart(near nominalStart: Int) -> Int? {
        let maxStart = rxBuffer.count - (preambleBits.count * symbolSamples)
        if maxStart <= 0 {
            return nil
        }
        let searchRadius = symbolSamples
        let from = max(0, nominalStart - searchRadius)
        let to = min(maxStart, nominalStart + searchRadius)
        if from > to {
            return nil
        }

        var bestStart = -1
        var bestScore = Int.min
        for candidate in from...to {
            let score = scorePreamble(at: candidate)
            if score > bestScore {
                bestScore = score
                bestStart = candidate
            }
        }
        return bestScore >= minPreambleScore ? bestStart : nil
    }

    private func decodeByte(at start: Int, startBit: Int) -> UInt8? {
        var value: UInt8 = 0
        for bitOffset in 0..<8 {
            guard let bit = decodeBit(at: start, bitIndex: startBit + bitOffset) else {
                return nil
            }
            value = (value << 1) | bit
        }
        return value
    }

    private func decodeBit(at start: Int, bitIndex: Int) -> UInt8? {
        let nominalStart = start + (bitIndex * symbolSamples)
        guard nominalStart >= 0 else {
            return nil
        }

        let searchFrom = max(0, nominalStart - symbolSearchRadius)
        let searchTo = min(rxBuffer.count - symbolSamples, nominalStart + symbolSearchRadius)
        if searchFrom > searchTo {
            return nil
        }

        var bestBit: UInt8?
        var bestScore: Float = -.greatestFiniteMagnitude
        var candidateStart = searchFrom
        while candidateStart <= searchTo {
            let result = classifyBit(windowStart: candidateStart)
            let strength = abs(result.highMag - result.lowMag)
            if strength > bestScore {
                bestScore = strength
                bestBit = result.bit
            }
            candidateStart += symbolSearchStep
        }
        return bestBit
    }

    private func classifyBit(windowStart: Int) -> (bit: UInt8, lowMag: Float, highMag: Float) {
        var lowI: Float = 0
        var lowQ: Float = 0
        var highI: Float = 0
        var highQ: Float = 0
        for idx in 0..<symbolSamples {
            let sample = rxBuffer[windowStart + idx]
            lowI += sample * lowRef.sin[idx]
            lowQ += sample * lowRef.cos[idx]
            highI += sample * highRef.sin[idx]
            highQ += sample * highRef.cos[idx]
        }
        let lowMag = sqrt((lowI * lowI) + (lowQ * lowQ))
        let highMag = sqrt((highI * highI) + (highQ * highQ))
        return (highMag > lowMag ? 1 : 0, lowMag, highMag)
    }

    private mutating func dropFront(_ count: Int) {
        if count <= 0 || rxBuffer.isEmpty {
            return
        }
        rxBuffer.removeFirst(min(count, rxBuffer.count))
    }

    private func windowRms(start: Int) -> Float {
        let end = start + symbolSamples
        if start < 0 || end > rxBuffer.count {
            return 0
        }
        var energy: Float = 0
        for index in start..<end {
            energy += rxBuffer[index] * rxBuffer[index]
        }
        return sqrt(energy / Float(symbolSamples))
    }

    private func majorityByte(_ bytes: [UInt8]) -> UInt8 {
        precondition(!bytes.isEmpty)
        var value: UInt8 = 0
        for bitIndex in 0..<8 {
            let mask = UInt8(1 << UInt8(7 - bitIndex))
            let ones = bytes.reduce(0) { count, byte in count + ((byte & mask) == 0 ? 0 : 1) }
            if ones * 2 >= bytes.count {
                value |= mask
            }
        }
        return value
    }

    private static func makeReference(
        freqHz: Int,
        sampleRateHz: Int,
        symbolSamples: Int,
        txGainCap: Float
    ) -> (sin: [Float], cos: [Float]) {
        let gain = txGainCap
        let values = (0..<symbolSamples).map { index -> (Float, Float) in
            let phase = (2.0 * Float.pi * Float(freqHz) * Float(index)) / Float(sampleRateHz)
            return (gain * Foundation.sin(phase), gain * Foundation.cos(phase))
        }
        return (values.map(\.0), values.map(\.1))
    }

    private static func byteToBits(_ value: UInt8) -> [UInt8] {
        (0..<8).map { bitIndex in (value >> UInt8(7 - bitIndex)) & 0x1 }
    }

    private static func crc8<S: Sequence>(_ bytes: S) -> Int where S.Element == UInt8 {
        var crc = 0
        for byte in bytes {
            crc ^= Int(byte)
        }
        return crc & 0xFF
    }
}
